File: scripts/item-generator/logging.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Ian Lucas. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { log } from "../utils.ts";

export function formatCount(count: number, singular: string, plural = `${singular}s`): string {
    return `${count} ${count === 1 ? singular : plural}`;
}

function formatDuration(start: number) {
    const elapsed = Date.now() - start;
    return elapsed < 1000 ? `${elapsed}ms` : `${(elapsed / 1000).toFixed(1)}s`;
}

export async function runStep(
    name: string,
    callback: () => Promise<void>,
    getSummary?: () => string | undefined
): Promise<void> {
    const start = Date.now();
    log(`${name}...`);
    await callback();
    const summary = getSummary?.();
    log(`${name} done${summary !== undefined ? ` (${summary})` : ""} in ${formatDuration(start)}.`);
}

File: scripts/item-generator/Logging.cs
using System.Diagnostics;

namespace ItemGenerator;

public static class Logging
{
    public static void Log(string message)
    {
        Console.WriteLine(message);
    }

    public static async Task RunStep(string name, Func<Task> callback, Func<string?>? getSummary = null)
    {
        var sw = Stopwatch.StartNew();
        Log($"{name}...");
        await callback();
        sw.Stop();
        var summary = getSummary?.Invoke();
        var duration = sw.ElapsedMilliseconds < 1000
            ? $"{sw.ElapsedMilliseconds}ms"
            : $"{sw.Elapsed.TotalSeconds:F1}s";
        Log($"{name} done{(summary != null ? $" ({summary})" : "")} in {duration}.");
    }

    public static string FormatCount(int count, string singular, string? plural = null)
    {
        plural ??= $"{singular}s";
        return $"{count} {(count == 1 ? singular : plural)}";
    }
}
